reconcile/acs_rbac.py

```diff
@@ -0,0 +1,568 @@
+                    role_group_mappings[group.role_name] = []
+                role_group_mappings[group.role_name].append(group)
+
+        # role and associated resources must be deleted in the proceeding order
+        for role in to_delete.values():
+            if not dry_run:
+                try:
+                    acs.delete_group_batch(role_group_mappings[role.name])
+                except Exception as e:
+                    logging.error(
+                        f"Failed to delete group(s) for role: {role.name}\t\n{e}"
+                    )
+                    continue
+            logging.info(
+                f"Deleted users from role '{role.name}': {[a.value for a in role.assignments]}"
+            )
+            # only delete rules associated with a system default roles
+            # do not continue to deletion of the role and associated access scope
+            if role.system_default:
+                continue
+            if not dry_run:
+                try:
+                    acs.delete_role(role.name)
+                except Exception as e:
+                    logging.error(f"Failed to delete role: {role.name}\t\n{e}")
+                    continue
+            logging.info(f"Deleted role '{role.name}'")
+            if not dry_run:
+                try:
+                    acs.delete_access_scope(access_scope_id_map[role.access_scope.name])
+                except Exception as e:
+                    logging.error(
+                        f"Failed to delete access scope for role: {role.name}\t\n{e}"
+                    )
+                    continue
+            logging.info(f"Deleted access scope '{role.access_scope.name}'")
+
+    def update_rbac(
+        self,
+        to_update: dict[str, DiffPair[AcsRole, AcsRole]],
+        acs: AcsApi,
+        auth_id: str,
+        dry_run: bool,
+    ) -> None:
+        """
+        Updates desired ACS roles as well as associated access scopes and rules
+
+        :param to_update: result of 'diff_iterables(current, desired).change' for ACS roles
+        :param acs: ACS api client
+        :param auth_id: id of auth provider within ACS instance to target for reconciliation
+        :param dry_run: run in dry-run mode
+        """
+        access_scope_id_map = {s.name: s.id for s in acs.get_access_scopes()}
+        permission_sets_id_map = {ps.name: ps.id for ps in acs.get_permission_sets()}
+        role_group_mappings: dict[str, dict[str, Group]] = {}
+        for group in acs.get_groups():
+            if group.role_name not in role_group_mappings:
+                role_group_mappings[group.role_name] = {}
+            role_group_mappings[group.role_name][group.value] = group
+
+        for role_diff_pair in to_update.values():
+            # auth rule (groups) portion
+            diff = diff_iterables(
+                role_diff_pair.current.assignments,
+                role_diff_pair.desired.assignments,
+                lambda x: x.value,
```

Comment:
Yes. In this case `key` corresponds to the attribute to specify for an auth rule (exs: `roleid`, `email`, `org_username`, etc. This integration will create rules using `org_username` as the key and value will be the unique username values.

Reply:
So you mean in `assignments`, `key` is always `org_username`, but `value` can be different username values, right? Then it looks good as long as `value` is unique across all items inside `assignments`. And based on this data structure, it can only be `add`, delete` and `identical`, no `change`.

Reply:
Correct. `org_username` corresponds to https://github.com/app-sre/qontract-schemas/blob/main/schemas/access/user-1.yml#L16